src/satosa/micro_services/primary_identifier.py

```diff
@@ -54,7 +54,7 @@ def constructPrimaryIdentifier(self, data, ordered_identifier_candidates):
 
             # Get the values asserted by the IdP for the configured list of attribute names for this candidate
             # and substitute None if the IdP did not assert any value for a configured attribute.
-            values = [ attributes.get(attribute_name, [None])[0] for attribute_name in candidate['attribute_names'] ]
+            values = [ attributes.get(attribute_name, [None])[0] for attribute_name in candidate['attribute_names'] if attribute_name != 'name_id' ]
```

Comment:
Please, correct me if I'm wrong.


The final value is a concatenation of the attribute-values of the candidate attributes. The candidates are a list of fallback-sets. `None` coming from `name_id` should not be an issue, as the configuration defines that _it is needed_ - if it is not the configuration should include all cases for the expected attributes.

The policy below says that if eptid is there, it should be used in combination with eppn. If it is not, then use eppn with NameID. Otherwise use eppn or NameID alone.

```yaml
ordered_identifier_candidates:
  - attribute_names:
      - eppn
      - eptid
  - attribute_names:
      - eppn
      - name_id
    name_id_format: urn:oasis:names:tc:SAML:2.0:nameid-format:persistent
  - attribute_names:
      - eppn
  - attribute_names:
      - name_id
    name_id_format: urn:oasis:names:tc:SAML:2.0:nameid-format:persistent
```

This means that:
- the combination of eppn&eptid is preferred (final value `<eppn><eptid>[<@scope>]`)
- if that is not available (eppn or eptid or both are not defined), then use the combination eppn&NameID when the name_id format is persistent (final value `<eppn><NameID>[<@scope>]`)
- if that is not available (eppn or name-id or both are not defined, or the name-id-format differs), then use eppn alone (final value `<eppn>[<@scope>]`)
- if that is not available (eppn is not defined), then use the name-id alone (final value `<NameID>[<@scope>]`)

---

In other words, all `attribute_names` are mandatory to have a value. With this change we say that we make an exception for name_id and it's now _optional_. This may make some use-cases easier to define (less fallback sets) but it will change the meaning of existing configurations and possibly break them (the cases that needed to match only when name_id had a value).

---

- Is this a configuration issue? and does the above help with setting up your configuration correctly?
- Have I missed something that should be addressed?

Let me know, I'm open to suggestions.

Reply:
Hi,

Thanks for the detailed review.

I agree that all `attribute_names` should be mandatory - and I'll try explaining why that still holds.

The key to that is that `name_id` is handled separately, around line 65 - in the block headed by `if 'name_id' in candidate['attribute_names']:`

If a matching NameID is found (a value is provided, and the required format is specified and matches the format actually provided), than the value (stored in `candidate_nameid_value`) is added into `values`.

On the other hand, if `name_id` is specified in the candidate's `attribute_names`, but a suitable value is not found, `candidate_nameid_value` is set to `None` - and this value is still added into `values`, invalidating the set produced by this `candidate` entry.  So the invariant is preserved.

The motivation for this change is that the expression this discussion is focused on (L57), when `candidate['attribute_names']` contains `name_id`, would *always* including a `None` in `values`, immediately marking it as invalid - because no value would ever be found.  Values for `name_id` are not present in `attributes`, they're only available as `data.subject_id` (value) + `data.subject_type` (NameIDFormat).

As this code already has special handling for `name_id` attribute names (as per above), this change only complements the special handling by excluding the attribute from the regular handling.

However, a `name_id` value is still mandatory - this is taken care of in the special handling block.

Hope this explains it.

Cheers,
Vlad

Reply:
yes, you are right! thanks for the explanation